{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perfect Numbers\n",
    "\n",
    "Determine if a number is perfect, abundant, or deficient based on\n",
    "Nicomachus' (60 - 120 CE) classification scheme for natural numbers.\n",
    "\n",
    "The Greek mathematician [Nicomachus](https://en.wikipedia.org/wiki/Nicomachus) devised a classification scheme for natural numbers, identifying each as belonging uniquely to the categories of **perfect**, **abundant**, or **deficient** based on their [aliquot sum](https://en.wikipedia.org/wiki/Aliquot_sum). The aliquot sum is defined as the sum of the factors of a number not including the number itself. For example, the aliquot sum of 15 is (1 + 3 + 5) = 9\n",
    "\n",
    "- **Perfect**: aliquot sum = number\n",
    "  - 6 is a perfect number because (1 + 2 + 3) = 6\n",
    "  - 28 is a perfect number because (1 + 2 + 4 + 7 + 14) = 28\n",
    "- **Abundant**: aliquot sum > number\n",
    "  - 12 is an abundant number because (1 + 2 + 3 + 4 + 6) = 16\n",
    "  - 24 is an abundant number because (1 + 2 + 3 + 4 + 6 + 8 + 12) = 36\n",
    "- **Deficient**: aliquot sum < number\n",
    "  - 8 is a deficient number because (1 + 2 + 4) = 7\n",
    "  - Prime numbers are deficient\n",
    "\n",
    "Implement a way to determine whether a given number is **perfect**. Depending on your language track, you may also need to implement a way to determine whether a given number is **abundant** or **deficient**.\n",
    "\n",
    "## Exception messages\n",
    "\n",
    "Sometimes it is necessary to raise an exception. When you do this, you should include a meaningful error message to\n",
    "indicate what the source of the error is. This makes your code more readable and helps significantly with debugging. Not\n",
    "every exercise will require you to raise an exception, but for those that do, the tests will only pass if you include\n",
    "a message.\n",
    "\n",
    "To raise a message with an exception, just write it as an argument to the exception type. For example, instead of\n",
    "`raise Exception`, you shold write:\n",
    "\n",
    "```python\n",
    "raise Exception(\"Meaningful message indicating the source of the error\")\n",
    "```\n",
    "\n",
    "## Submitting Exercises\n",
    "\n",
    "Note that, when trying to submit an exercise, make sure the solution is in the `$EXERCISM_WORKSPACE/python/perfect-numbers` directory.\n",
    "\n",
    "You can find your Exercism workspace by running `exercism debug` and looking for the line that starts with `Workspace`.\n",
    "\n",
    "For more detailed information about running tests, code style and linting,\n",
    "please see the [help page](http://exercism.io/languages/python).\n",
    "\n",
    "## Source\n",
    "\n",
    "Taken from Chapter 2 of Functional Thinking by Neal Ford. [http://shop.oreilly.com/product/0636920029687.do](http://shop.oreilly.com/product/0636920029687.do)\n",
    "\n",
    "## Submitting Incomplete Solutions\n",
    "It's possible to submit an incomplete solution so you can see how others have completed the exercise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify(number):\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "\n",
    "\n",
    "\n",
    "# Tests adapted from `problem-specifications//canonical-data.json` @ v1.1.0\n",
    "\n",
    "class PerfectNumbersTest(unittest.TestCase):\n",
    "    def test_smallest_perfect_number(self):\n",
    "        self.assertIs(classify(6), \"perfect\")\n",
    "\n",
    "    def test_medium_perfect_number(self):\n",
    "        self.assertIs(classify(28), \"perfect\")\n",
    "\n",
    "    def test_large_perfect_number(self):\n",
    "        self.assertIs(classify(33550336), \"perfect\")\n",
    "\n",
    "\n",
    "class AbundantNumbersTest(unittest.TestCase):\n",
    "    def test_smallest_abundant_number(self):\n",
    "        self.assertIs(classify(12), \"abundant\")\n",
    "\n",
    "    def test_medium_abundant_number(self):\n",
    "        self.assertIs(classify(30), \"abundant\")\n",
    "\n",
    "    def test_large_abundant_number(self):\n",
    "        self.assertIs(classify(33550335), \"abundant\")\n",
    "\n",
    "\n",
    "class DeficientNumbersTest(unittest.TestCase):\n",
    "    def test_smallest_prime_deficient_number(self):\n",
    "        self.assertIs(classify(2), \"deficient\")\n",
    "\n",
    "    def test_smallest_nonprime_deficient_number(self):\n",
    "        self.assertIs(classify(4), \"deficient\")\n",
    "\n",
    "    def test_medium_deficient_number(self):\n",
    "        self.assertIs(classify(32), \"deficient\")\n",
    "\n",
    "    def test_large_deficient_number(self):\n",
    "        self.assertIs(classify(33550337), \"deficient\")\n",
    "\n",
    "    def test_edge_case(self):\n",
    "        self.assertIs(classify(1), \"deficient\")\n",
    "\n",
    "\n",
    "class InvalidInputsTest(unittest.TestCase):\n",
    "    def test_zero(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            classify(0)\n",
    "\n",
    "    def test_negative(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            classify(-1)\n",
    "\n",
    "    # Utility functions\n",
    "    def setUp(self):\n",
    "        try:\n",
    "            self.assertRaisesRegex\n",
    "        except AttributeError:\n",
    "            self.assertRaisesRegex = self.assertRaisesRegexp\n",
    "\n",
    "    def assertRaisesWithMessage(self, exception):\n",
    "        return self.assertRaisesRegex(exception, r\".+\")\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
